File: server/src/main/java/com/airship/game/game/Bullet.java
package com.airship.game.game;

public class Bullet {

    private String id;
    private volatile float x;
    private volatile float y;
    private float vx;
    private float vy;
    private String ownerId;
    private int damage;
    private String color;
    private float radius = 4;
    private boolean active = true;
    private long age = 0;
    private long maxAge = 2000;

    private static int bulletCounter = 0;

    public Bullet(float x, float y, float vx, float vy, String ownerId, int damage, String color) {
        this.id = "bullet_" + (++bulletCounter);
        this.x = x;
        this.y = y;
        this.vx = vx;
        this.vy = vy;
        this.ownerId = ownerId;
        this.damage = damage;
        this.color = color;
    }

    public void update(float deltaTime) {
        x += vx * deltaTime;
        y += vy * deltaTime;
        age += deltaTime * 1000;

        if (age >= maxAge) {
            active = false;
        }

        if (x < -20 || x > 980 || y < -20 || y > 660) {
            active = false;
        }
    }

    public boolean checkCollision(Player player) {
        if (!active || player == null || !player.isActive()) return false;
        if (player.getId().equals(ownerId)) return false;
        if (player.isInvincible()) return false;

        float playerRadius = player.getWidth() / 2.0f * 0.7f;
        float dx = x - (player.getX() + player.getWidth() / 2.0f);
        float dy = y - (player.getY() + player.getHeight() / 2.0f);
        float dist = (float) Math.sqrt(dx * dx + dy * dy);

        return dist < playerRadius + radius;
    }

    public String getId() { return id; }
    public float getX() { return x; }
    public float getY() { return y; }
    public float getVx() { return vx; }
    public float getVy() { return vy; }
    public String getOwnerId() { return ownerId; }
    public int getDamage() { return damage; }
    public String getColor() { return color; }
    public float getRadius() { return radius; }
    public boolean isActive() { return active; }

    public void setActive(boolean active) {
        this.active = active;
    }
}

File: client/js/bullet.js
class Bullet {
    constructor(x, y, vx, vy, ownerId, damage = 10, color = '#ff6600') {
        this.x = x;
        this.y = y;
        this.vx = vx;
        this.vy = vy;
        this.ownerId = ownerId;
        this.damage = damage;
        this.color = color;
        this.radius = 4;
        this.active = true;
        this.age = 0;
        this.maxAge = 2000;
    }

    update(deltaTime) {
        this.x += this.vx * deltaTime;
        this.y += this.vy * deltaTime;
        this.age += deltaTime;
        
        if (this.age >= this.maxAge) {
            this.active = false;
        }

        if (this.x < -20 || this.x > 980 || this.y < -20 || this.y > 660) {
            this.active = false;
        }
    }

    draw(ctx) {
        ctx.save();
        ctx.fillStyle = this.color;
        ctx.shadowColor = this.color;
        ctx.shadowBlur = 8;
        
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.radius, 0, Math.PI * 2);
        ctx.fill();
        
        ctx.fillStyle = '#ffffff';
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.radius * 0.4, 0, Math.PI * 2);
        ctx.fill();
        
        ctx.restore();
    }

    toData() {
        return {
            x: this.x,
            y: this.y,
            vx: this.vx,
            vy: this.vy,
            ownerId: this.ownerId,
            damage: this.damage,
            color: this.color,
            radius: this.radius
        };
    }

    static fromData(data) {
        const bullet = new Bullet(
            data.x, data.y, data.vx, data.vy,
            data.ownerId, data.damage, data.color
        );
        if (data.radius) bullet.radius = data.radius;
        return bullet;
    }
}
